pootle/apps/pootle_store/syncer.py

```diff
@@ -0,0 +1,363 @@
+            u"Converting %s to %s",
+            self.store.pootle_path,
+            fileclass)
+        output = fileclass()
+        output.settargetlanguage(self.language.code)
+        # FIXME: we should add some headers
+        for unit in self.store.units.iterator():
+            output.addunit(
+                self.unit_sync_class(unit).convert(output.UnitClass))
+        return output
+
+    def _getclass(self, obj):
+        try:
+            return getclass(obj)
+        except ValueError:
+            raise ValueError(
+                "Unable to find conversion class for Store '%s'"
+                % self.store.name)
+
+    def get_new_units(self, old_ids, new_ids):
+        return self.store.findid_bulk(
+            [self.store.dbid_index.get(uid)
+             for uid
+             in new_ids - old_ids])
+
+    def get_obsolete_units(self, old_ids, new_ids):
+        return (
+            self.disk_store.findid(uid)
+            for uid
+            in old_ids - new_ids)
+
+    def obsolete_unit(self, unit, conservative):
+        obsolete = 0
+        deleted = 0
+        del_unit = False
+        if not unit.istranslated():
+            del_unit = True
+        elif not conservative:
+            obsolete = 1
+            unit.makeobsolete()
+            if not unit.isobsolete():
+                del_unit = True
+        if del_unit:
+            deleted = 1
+            del unit
+        return obsolete, deleted
+
+    def update_structure(self, obsolete_units, new_units, conservative):
+        obsolete = 0
+        deleted = 0
+        added = 0
+        for unit in obsolete_units:
+            _obsolete, _deleted = self.obsolete_unit(unit, conservative)
+            if _obsolete:
+                obsolete += 1
+            if _deleted:
+                deleted += 1
+        for unit in new_units:
+            newunit = unit.convert(self.disk_store.UnitClass)
+            self.disk_store.addunit(newunit)
+            added += 1
+        return obsolete, deleted, added
+
+    def create_store_file(self, last_revision, user):
+        logging.debug(u"Creating file %s", self.store.pootle_path)
+        store = self.convert()
+        store.savefile(self.store_file_path)
+        log(u"Created file for %s [revision: %d]" %
+            (self.store.pootle_path, last_revision))
+        self.store.file = self.store_file_path
+        self.update_store_header(user=user)
+        self.store.file.savestore()
+        self.store.file_mtime = self.store.get_file_mtime()
+        self.store.last_sync_revision = last_revision
+        self.store.save()
+
+    def update_newer(self, last_revision):
+        return (
+            not self.store.file.exists()
+            or (last_revision >= self.store.last_sync_revision))
+
+    def get_dbid_index(self, update=False, obsolete=False):
+        """build a quick mapping index between unit ids and database ids"""
+        if update or not hasattr(self.store, "dbid_index"):
```

Comment:
afaict `store.dbid_index` was defined by `require_dbid_index`. this PR drops `require_dbid_index` method. I wonder if   `hasattr(self.store, "dbid_index")` condition is still necessary here.